### PR TITLE
Improve the performance of TextSource by reducing how many byte[]s are copied (fixes #23193)

### LINKED ISSUE #23193: [Feature Request]: Reduce number of byte[] copies in TextSource
### What would you like to happen?

The current TextSource implementation is spending a lot of time during `byte[]` copying:
TextSource old implementation performance in pipeline

Hadoop `LineReader.java` implementation is signficantly faster (~2x) when handling typical files due to an implementation that reduces how many `byte[]`s are copied. A simple benchmark reading 10 million lines (60-120 characters long) shows that it takes about ~2.05 seconds to process such a file while the Apache Beam TextSource takes ~4.03 seconds.

### Issue Priority

Priority: 2

### Issue Component

Component: io-java-text

### PR DESCRIPTION
**Pull request by lukecwik**
This makes TextSource take about 2.3x less CPU resources during decoding.

Before this change:
```
TextSourceBenchmark.benchmarkTextSource        thrpt    5  0.248 ± 0.029  ops/s
```

After this change:
```
TextSourceBenchmark.benchmarkHadoopLineReader  thrpt    5  0.465 ± 0.064  ops/s
TextSourceBenchmark.benchmarkTextSource        thrpt    5  0.575 ± 0.059  ops/s
```

Fixes #23193

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @bhisevishal

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Java_Examples_Dataflow PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java [hunk 1]
```diff
@@ -152,131 +160,260 @@ protected void startReading(ReadableByteChannel channel) throws IOException {
...
-     * till either EOF or the delimiter bounds are found.
+     * Loosely based upon <a
+     * href="https://github.com/hanborq/hadoop/blob/master/src/core/org/apache/hadoop/util/LineReader.java">Hadoop
+     * LineReader.java</a>
      *
-     * <p>This fills the buffer and updates the positions as follows:
+     * <p>We're reading data from inChannel, but the head of the stream may be already buffered in
+     * buffer, so we have several cases: 1. No newline characters are in the buffer, so we need to
+     * copy everything and read another buffer from the stream. 2. An unambiguously terminated line
+     * is in buffer, so we just create currentValue 3. Ambiguously terminated line is in buffer,
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java:235]: nit: could this be an HTML ordered list?

**Review by TheNeuralBit — APPROVED** [review]
LGTM

Please link the issue by adding "Fixes #23193" in the description.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java (hunk 1 above)
Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java:235]: Done

**Comment by lukecwik** [comment]
> LGTM
> 
> Please link the issue by adding "Fixes #23193" in the description.

Its already part of the description.

**Comment by TheNeuralBit** [comment]
> > LGTM
> > Please link the issue by adding "Fixes #23193" in the description.
> 
> Its already part of the description, what am I missing?

You have it in the title, but not in the description. There's some documentation here: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Comment by lukecwik** [comment]
> > > LGTM
> > > Please link the issue by adding "Fixes #23193" in the description.
> > 
> > 
> > Its already part of the description, what am I missing?
> 
> You have it in the title, but not in the description. There's some documentation here: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

That is an annoying limitation. Thanks for pointing it out.

**Comment by lukecwik** [comment]
Run Spotless PreCommit

**Comment by lukecwik** [comment]
Run Java PreCommit

**Pull request merged**